consensus/src/dag/bootstrap.rs

```diff
@@ -1,136 +1,303 @@
+        let validators = self.epoch_state.verifier.get_ordered_account_addresses();
+
+        // A backoff policy that starts at 100ms and doubles each iteration.
+        let rb_backoff_policy = ExponentialBackoff::from_millis(2).factor(50);
+        let rb = Arc::new(ReliableBroadcast::new(
+            validators.clone(),
+            self.rb_network_sender.clone(),
+            rb_backoff_policy,
+            self.time_service.clone(),
+            // TODO: add to config
+            Duration::from_millis(500),
+        ));
+
+        let (dag_fetcher, fetch_requester, node_fetch_waiter, certified_node_fetch_waiter) =
+            DagFetcherService::new(
+                self.epoch_state.clone(),
+                self.dag_network_sender.clone(),
+                dag.clone(),
+                self.time_service.clone(),
+            );
+        let fetch_requester = Arc::new(fetch_requester);
+
+        let dag_driver = DagDriver::new(
+            self.self_peer,
+            self.epoch_state.clone(),
+            dag.clone(),
+            self.payload_client.clone(),
+            rb,
+            self.time_service.clone(),
+            self.storage.clone(),
+            order_rule,
+            fetch_requester.clone(),
+            self.db.clone(),
+        );
+        let rb_handler = NodeBroadcastHandler::new(
             dag.clone(),
-            time_service.clone(),
+            self.signer.clone(),
+            self.epoch_state.clone(),
+            self.storage.clone(),
+            fetch_requester,
+        );
+        let fetch_handler = FetchRequestHandler::new(dag, self.epoch_state.clone());
+
+        let dag_handler = NetworkHandler::new(
+            self.epoch_state.clone(),
+            rb_handler,
+            dag_driver,
+            fetch_handler,
+            node_fetch_waiter,
+            certified_node_fetch_waiter,
+            state_sync_trigger,
+        );
+
+        (dag_handler, dag_fetcher)
+    }
+
+    async fn bootstrapper(
+        self,
+        mut dag_rpc_rx: Receiver<Author, IncomingDAGRequest>,
+        ordered_nodes_tx: UnboundedSender<OrderedBlocks>,
+        shutdown_rx: oneshot::Receiver<()>,
+    ) -> anyhow::Result<()> {
+        let adapter = Arc::new(NotificationAdapter::new(
+            ordered_nodes_tx,
+            self.storage.clone(),
+        ));
+
+        let sync_manager = DagStateSynchronizer::new(
+            self.epoch_state.clone(),
+            adapter.clone(),
+            self.time_service.clone(),
+            self.state_computer.clone(),
+            self.storage.clone(),
         );
-    let fetch_requester = Arc::new(fetch_requester);
 
-    let dag_driver = DagDriver::new(
+        // TODO: fetch the correct block info
+        let ledger_info = LedgerInfoWithSignatures::new(
+            LedgerInfo::new(BlockInfo::empty(), HashValue::zero()),
+            AggregateSignature::empty(),
+        );
+
+        let mut shutdown_rx = shutdown_rx.into_stream();
```

Comment:
I don't think you need this, just &mut shutdown_rx is enough

Reply:
I need this btw, because the oneshot::Receiver seems moved within the select statement.